### PR TITLE
ISPN-14423 Improve configuration parser error reporting

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-14423

ISPN-14423 Improve configuration parser error reporting
    
* Report location of parsing error and, if possible, also the file name
* Report the full attribute path, eg. `local-cache.encoding.key.media-type`
* Show valid values for a wrong attribute (enums, etc)
* Start simplifying parsing of attributeSets
* Support YAML boolean types: `true|false|yes|no|on|off|y|n`
* Add messages to all error responses in the REST API

Notes for reviewers:

* Many changes are about adding an `attributes()` method to builders 
* Many changes are about converting the legacy `memory` configuration, deprecated in 11.0, to the new style
* The interesting stuff lies in the `org.infinispan.commons.configuration` package and the parsers

### CODE REVIEW

File: cli/src/main/java/org/infinispan/cli/commands/Config.java [hunk 1]
```diff
@@ -156,10 +160,19 @@ public boolean isHelp() {
...
+               is = System.in;
+               resolver = ConfigurationResourceResolver.DEFAULT;
+            } else {
+               File file = new File(input.getAbsolutePath());
+               is = new FileInputStream(file);
+               resolver = new URLConfigurationResourceResolver(file.toURI().toURL());
+            }
             is = input == null ? System.in : new FileInputStream(input.getAbsolutePath());
-            ConfigurationBuilderHolder holder = registry.parse(is, null, null); // Auto-detect type
+            ConfigurationBuilderHolder holder = registry.parse(is, resolver, null); // Auto-detect type
```
Comment (ryanemerson) [inline cli/src/main/java/org/infinispan/cli/commands/Config.java:175]: `is` is being assigned again after the if/else.

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java [hunk 2]
```diff
@@ -82,6 +90,14 @@ public boolean isGlobal() {
       return global;
    }
 
+   public boolean isDeprecated() {
+      return (deprecatedMajor < Integer.MAX_VALUE);
+   }
+
+   public boolean isDeprecated(int major, int minor) {
```
Comment (ryanemerson) [inline commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java:97]: Neither `isDeprecated` method appears to be called. Do you have ause-case in mind for these?

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java [hunk 3]
```diff
@@ -82,6 +90,14 @@ public boolean isGlobal() {
       return global;
    }
 
+   public boolean isDeprecated() {
+      return (deprecatedMajor < Integer.MAX_VALUE);
```
Comment (ryanemerson) [inline commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java:94]: :thinking:

File: core/src/main/java/org/infinispan/configuration/parsing/ParseUtils.java [hunk 4]
```diff
@@ -5,356 +5,421 @@
...
-     */
-    public static ConfigurationReaderException missingRequired(final ConfigurationReader reader, final Set<?> required) {
-        final StringBuilder b = new StringBuilder();
-        Iterator<?> iterator = required.iterator();
-        while (iterator.hasNext()) {
-            final Object o = iterator.next();
-            b.append(o.toString());
-            if (iterator.hasNext()) {
-                b.append(", ");
+   private ParseUtils() {
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/configuration/parsing/ParseUtils.java:24]: Can you revert the format changes here? It makes it difficult to review and effectively means we lose the git history.

File: core/src/test/java/org/infinispan/configuration/UpdatableConfigurationTest.java [hunk 5]
```diff
@@ -20,10 +23,32 @@ public void testUpdateConfiguration() {
...
       builder.expiration().maxIdle(3000); // Lifespan will be set to the default
-      configuration.update(builder.build());
+      configuration.update(null, builder.build());
 
       // Check that only the modified attributes have been updated
       assertEquals(-1, configuration.expiration().lifespan());
       assertEquals(3000, configuration.expiration().maxIdle());
    }
+
+   //@Test(expectedExceptions = IllegalArgumentException.class)
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/configuration/UpdatableConfigurationTest.java:33]: Remove.

**Review by ryanemerson — COMMENTED** [review]
Good use of `ParserUtils` to remove lots of repetitive code :slightly_smiling_face:

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java (hunk 3 above)
Comment (tristantarrant) [reply commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java:94]: By default, attributes are non-deprecated and have `deprecatedMajor = Integer.MAX_VALUE`.

File: commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java (hunk 2 above)
Comment (tristantarrant) [reply commons/all/src/main/java/org/infinispan/commons/configuration/attributes/AttributeDefinition.java:93]: I've fixed `ParseUtils.parseAttributes()` to report use of deprecated attributes. Now we just need to decorate those we want to deprecate :)

File: core/src/main/java/org/infinispan/configuration/parsing/ParseUtils.java (hunk 4 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/configuration/parsing/ParseUtils.java]: Done. Unfortunately this file use 4-space indents...

File: core/src/test/java/org/infinispan/configuration/UpdatableConfigurationTest.java (hunk 5 above)
Comment (tristantarrant) [reply core/src/test/java/org/infinispan/configuration/UpdatableConfigurationTest.java]: Done

File: cli/src/main/java/org/infinispan/cli/commands/Config.java (hunk 1 above)
Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/commands/Config.java:174]: Fixed

**Comment by tristantarrant** [comment]
I have split the PR into multiple commits

**Comment by ryanemerson** [comment]
The `*IT` test failures seem related.

**Comment by ryanemerson** [comment]
Now there's new failures that are related :slightly_smiling_face:

**Comment by tristantarrant** [comment]
Fixed everything

**Comment by ryanemerson** [comment]
Thanks @tristantarrant

**Pull request merged**